Modernize Instructor Question Settings page

mylesw27 opened the pull request:
This PR intends to convert the instructorQuestionSettings page from JavaScript and EJS to TypeScript and HTML template literals. The goal is to modernize this page first and subsequent PRs will work on the UI editor (i.e. adding a form and editing question setting in browser).

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,584 @@
+                                  : html`
+                                      Shared With:
+                                      ${sharingSetsIn.map(function (sharing_set) {
+                                        return html`<span class="badge color-gray1"
+                                          >${sharing_set?.name}</span
+                                        >`;
```
Usual comment about splitting this into multiple lines.

mylesw27 replied:
This is now split out to multiple lines (now lines 220-224) in [e8cc26c](https://github.com/PrairieLearn/PrairieLearn/pull/9571/commits/e8cc26c79af51815e4e9bdbf5d82f9f432f5decc).

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,584 @@
+  qids: (string | null)[];
+  a_with_q_for_all_ci: SelectedAssessments[];
+  sharingEnabled: boolean;
+  sharingSetsIn: (SharingSet | null)[];
```
When would we expect this array to contain `null` items? Did you mean this instead?

```suggestion
  sharingSetsIn: SharingSet[] | null;
```

That said, I expect we could simplify this just just always passing in an array, and just making sure it's empty if sharing isn't enabled.

mylesw27 replied:
This array should not contain null items and should alway pass an array. Removed `null` in [e8cc26c](https://github.com/PrairieLearn/PrairieLearn/pull/9571/commits/e8cc26c79af51815e4e9bdbf5d82f9f432f5decc).

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:
```diff
@@ -222,56 +213,69 @@ router.get(
-      let result = await sqldb.queryAsync(sql.select_sharing_sets, {
+    let sharingSetsIn, sharingSetsOther;
+    if (sharingEnabled) {
+      const result = await sqldb.queryAsync(sql.select_sharing_sets, {
```
Consider using Zod validation here.

mylesw27 replied:
Updated to use `queryRows` instead of `queryAsync` and validating with `SharingSetSchema` in [e8cc26c](https://github.com/PrairieLearn/PrairieLearn/pull/9571/commits/e8cc26c79af51815e4e9bdbf5d82f9f432f5decc).

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,584 @@
+  resLocals: Record<string, any>;
+  questionTestPath: string;
+  questionTestCsrfToken: string;
+  isEnterprise: boolean;
```
Rather than passing this in from the route handler, consider just directly invoking `isEnterprise()` here.

mylesw27 replied:
Removed call for `inEnterprise()` from the route handler and moved the call inline where it is used in the `PubliclyShareModal` in `instructorQuestionSettings.html.ts`.

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,584 @@
+              </tbody>
+            </table>
+
+            <!--
+          We only want to show the card footer if it will have something inside. Since it is easy
+          to make a mistake when merging logical conditions, we will simply "or" them all.
+        -->
```
Is this comment necessary?

mylesw27 replied:
Removed in [e8cc26c](https://github.com/PrairieLearn/PrairieLearn/pull/9571/commits/e8cc26c79af51815e4e9bdbf5d82f9f432f5decc).

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,584 @@
+  `.toString();
+}
+
+function copyForm({
```
As with other pages, these component-ish functions should have `PascalCase` names.

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,584 @@
+}: {
+  csrfToken: string;
+  exampleCourse: boolean;
+  editableCourses: Record<string, any>[];
```
I believe you can use a more precise type than this?

mylesw27 replied:
Yes, sorry used this during development and forgot to actually put a specific type in. Now using type `CourseWithPermissions[]` in [e8cc26c](https://github.com/PrairieLearn/PrairieLearn/pull/9571/commits/e8cc26c79af51815e4e9bdbf5d82f9f432f5decc).

nwalters512 reviewed (commented):
It seems like you didn't fully test this page? The "Delete question" button doesn't work, and the "cancel" button on the "Make a copy of this question" popover doesn't do anything.

mylesw27 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,584 @@
+          We only want to show the card footer if it will have something inside. Since it is easy
+          to make a mistake when merging logical conditions, we will simply "or" them all.
+        -->
+            ${(editableCourses &&
```
Removed in [e8cc26c](https://github.com/PrairieLearn/PrairieLearn/pull/9571/commits/e8cc26c79af51815e4e9bdbf5d82f9f432f5decc).

mylesw27 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,584 @@
+  isEnterprise,
+  questionGHLink,
+  qids,
+  a_with_q_for_all_ci,
```
Updated name to `assessmentsWithQuestion` in [e8cc26c](https://github.com/PrairieLearn/PrairieLearn/pull/9571/commits/e8cc26c79af51815e4e9bdbf5d82f9f432f5decc).

mylesw27 commented:
> It seems like you didn't fully test this page? The "Delete question" button doesn't work, and the "cancel" button on the "Make a copy of this question" popover doesn't do anything.

Sorry, had fully tested locally and then made changes to pass CI and forgot to actually test the page again. I had changed modal IDs to satisfy CI and it broke the modal button behavior. Now updated and all buttons should be working as of [e8cc26c](https://github.com/PrairieLearn/PrairieLearn/pull/9571/commits/e8cc26c79af51815e4e9bdbf5d82f9f432f5decc).

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,566 @@
+          )}
+          <div class="card mb-4">
+            <div class="card-header bg-primary text-white d-flex">
+              Question TESTING${resLocals.question.qid}
```
```suggestion
              Question ${resLocals.question.qid}
```

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,566 @@
+        Are you sure you want to delete the question
+        <strong>${qid}</strong>?
+      </p>
+      ${assessmentsWithQuestion
```
An empty array is still truthy in JavaScript. You want this instead:

```suggestion
      ${assessmentsWithQuestion.length
```

nwalters512 reviewed (commented):
A few more issues I caught during local testing.

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,565 @@
+  questionTestPath: string;
+  questionTestCsrfToken: string;
+  questionGHLink: string | null;
+  qids: (string | null)[];
```
When will this array contain null elements?

mylesw27 replied:
Removed `null` in [e66ab28](https://github.com/PrairieLearn/PrairieLearn/pull/9571/commits/e66ab286e419ca64e565b7f65a79937d66b73331).

nwalters512 replied:
Seems like you also need to update a corresponding Zod type.

mylesw27 replied:
Good catch. Updated Zod type in [859ec79](https://github.com/PrairieLearn/PrairieLearn/pull/9571/commits/859ec79031dbc9c954fdada6e61b53a6dc00f3f7).

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,565 @@
+                        `
+                      : ''}
+                    ${questionGHLink
+                      ? html` <a target="_blank" href="${questionGHLink}"> view on GitHub </a> `
```
```suggestion
                      ? html`<a target="_blank" href="${questionGHLink}"> view on GitHub </a>`
```

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,565 @@
+                                  <span>Edit</span>
+                                </a>
+                              `
+                            : html`info.json`}
```
We end up rendering `info.json` twice in this case:

<img width="291" alt="Screenshot 2024-04-01 at 15 22 56" src="https://github.com/PrairieLearn/PrairieLearn/assets/1476544/ae791051-5140-458e-936d-6ab4b2cd656a">

I think we should just render nothing (`''`) here in this branch.

mylesw27 replied:
Updated to render `''` here in [d5b67e6](https://github.com/PrairieLearn/PrairieLearn/pull/9571/commits/d5b67e64989f49d989864e2afd948f2db1b57803).

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,565 @@
+      <input type="hidden" name="__action" value="copy_question" />
+      <input type="hidden" name="__csrf_token" value="${csrfToken}" />
+      <div class="form-group">
+        <label for="to-course-id-select"
+          >The copied question will be added to the following course:</label
+        >
```
```suggestion
        <label for="to-course-id-select">
          The copied question will be added to the following course:
        </label>
```

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:
```diff
@@ -0,0 +1,565 @@
+        >
+        <select class="form-control" id="to-course-id-select" name="to_course_id" required>
+          ${exampleCourse
+            ? html` <option hidden disabled selected value>-- select a course --</option> `
```
```suggestion
            ? html`<option hidden disabled selected value>-- select a course --</option>`
```

nwalters512 reviewed (approved):
A few final nits; the main one to address before merging is the double-rendering of `info.json` in some cases.

nwalters512 reviewed (approved):
Approval stands!
